test/test_inverse.py

```diff
@@ -654,3 +655,102 @@ def test_KFAC_inverse_damped_torch_matvec(
 
     # Test against _matmat
     report_nonclose(inv_KFAC @ x.cpu().numpy(), inv_KFAC_x.cpu().numpy())
+
+
+def compare_state_dicts(state_dict: dict, state_dict_new: dict):
+    """Compare two state dicts recursively."""
+    assert len(state_dict) == len(state_dict_new)
+    for value, value_new in zip(state_dict.values(), state_dict_new.values()):
+        if isinstance(value, torch.Tensor):
+            assert torch.allclose(value, value_new)
+        elif isinstance(value, dict):
+            compare_state_dicts(value, value_new)
+        elif isinstance(value, tuple):
+            assert len(value) == len(value_new)
+            assert all(isinstance(v, type(v2)) for v, v2 in zip(value, value_new))
+            assert all(
+                torch.allclose(torch.as_tensor(v), torch.as_tensor(v2))
+                for v, v2 in zip(value, value_new)
+            )
+        else:
+            assert value == value_new
+
+
+def test_KFAC_inverse_save_and_load_state_dict():
+    """Test that KFACInverseLinearOperator can be saved and loaded from state dict."""
+    torch.manual_seed(0)
+    batch_size, D_in, D_out = 4, 3, 2
+    X = torch.rand(batch_size, D_in)
+    y = torch.rand(batch_size, D_out)
+    model = torch.nn.Linear(D_in, D_out)
+
+    params = list(model.parameters())
+    # create and compute KFAC
+    kfac = KFACLinearOperator(
+        model,
+        MSELoss(reduction="sum"),
+        params,
+        [(X, y)],
+        loss_average=None,
+    )
+
+    # create inverse KFAC
+    inv_kfac = KFACInverseLinearOperator(
+        kfac, damping=1e-2, use_heuristic_damping=True, retry_double_precision=False
+    )
+    _ = inv_kfac @ eye(kfac.shape[1])  # to trigger inverse computation
+
+    # save state dict
+    state_dict = inv_kfac.state_dict()
+    torch.save(state_dict, "inv_kfac_state_dict.pt")
+
+    # create new inverse KFAC with different linop input and try to load state dict
+    wrong_kfac = KFACLinearOperator(model, CrossEntropyLoss(), params, [(X, y)])
+    inv_kfac_wrong = KFACInverseLinearOperator(wrong_kfac)
+    with raises(ValueError, match="mismatch"):
+        inv_kfac_wrong.load_state_dict(torch.load("inv_kfac_state_dict.pt"))
+
+    # create new inverse KFAC and load state dict
+    inv_kfac_new = KFACInverseLinearOperator(kfac)
+    inv_kfac_new.load_state_dict(torch.load("inv_kfac_state_dict.pt"))
+
+    # check that the two inverse KFACs are equal
+    compare_state_dicts(inv_kfac.state_dict(), inv_kfac_new.state_dict())
+    test_vec = torch.rand(inv_kfac.shape[1])
+    report_nonclose(inv_kfac @ test_vec, inv_kfac_new @ test_vec)
```

Comment:
Still think it would be good to compute the first matvec before calling `inv_kfac.state_dict()`. Lmk what you think.

Reply:
Sorry, forgot to comment on this: I first changed it, but then realized when running the tests that there was a reason for this order -- if we call the matrix vector product first, it will just compute KFAC from scratch for the new linear operator and since params, data, etc match the result will always be the same.